### PR TITLE
MM-20540: Migrate 'components/user_settings/advanced' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR migrates 'components/user_settings/advanced' module and associated tests to TypeScript

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/13701
JIRA: https://mattermost.atlassian.net/browse/MM-20540

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: components/user_settings/advanced/join_leave_section/join_leave_section.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ type Props = {
     currentUserId: string;
     joinLeave?: string;
     onUpdateSection: (section?: string) => void;
-    renderOnOffLabel: (label: string) => string;
+    renderOnOffLabel: (label: string) => string | JSX.Element;
```
Comment (hmhealey) [inline components/user_settings/advanced/join_leave_section/join_leave_section.tsx:20]: `React.ReactNode` covers everything renderable (strings, JSX, null, etc), and it ends up being a bit shorter here
```suggestion
    renderOnOffLabel: (label: string) => React.ReactNode;
```

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 2]
```diff
@@ -448,7 +476,7 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
         );
     }
 
-    renderFeatureLabel(feature) {
+    renderFeatureLabel(feature: string): JSX.Element | null {
```
Comment (hmhealey) [inline components/user_settings/advanced/user_settings_advanced.tsx:479]: Similarly, most of these can be `React.ReactNode` as well

File: package-lock.json [hunk 3]
```diff
@@ -12,9 +12,9 @@
         "packages/components"
       ],
       "dependencies": {
-        "@floating-ui/react-dom": "0.7.1",
+        "@floating-ui/react-dom": "0.7.2",
```
Comment (hmhealey) [inline package-lock.json:15]: Would you mind reverting the changes in this file? I've submitted a separate PR to address them

https://github.com/mattermost/mattermost-webapp/pull/10832

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 3 above)
Comment (sinansonmez) [reply package-lock.json:15]: Sorry my mistake. I wasnt planning to commit these changes. I mistakenly committed this

File: components/user_settings/advanced/join_leave_section/join_leave_section.tsx (hunk 1 above)
Comment (sinansonmez) [reply components/user_settings/advanced/join_leave_section/join_leave_section.tsx]: Thanks for the suggestion, updated

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 2 above)
Comment (sinansonmez) [reply components/user_settings/advanced/user_settings_advanced.tsx]: Yep, you are right. Updated

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 4]
```diff
@@ -93,16 +119,18 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
...
         const settings = this.state.settings;
         settings[setting] = value;
-        this.setState(settings);
+
+        this.setState((prevState) => ({...prevState, ...settings}));
     }
 
-    toggleFeature = (feature, checked) => {
+    toggleFeature = (feature: string, checked: boolean): void => {
         const settings = this.state.settings;
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:134]: nit
```suggestion
        const {settings} = this.state;
```

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 5]
```diff
@@ -168,19 +196,19 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
...
 
-    handleHideDeactivateAccountModal = () => {
+    handleHideDeactivateAccountModal = (): void => {
         this.setState({
             showDeactivateAccountModal: false,
         });
     }
 
-    handleUpdateSection = (section) => {
+    handleUpdateSection = (section: string | undefined): void => {
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:211]:
```suggestion
    handleUpdateSection = (section?: string) => {
```

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 6]
```diff
@@ -168,19 +196,19 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
...
 
-    handleShowDeactivateAccountModal = () => {
+    handleShowDeactivateAccountModal = (): void => {
         this.setState({
             showDeactivateAccountModal: true,
         });
     }
 
-    handleHideDeactivateAccountModal = () => {
+    handleHideDeactivateAccountModal = (): void => {
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:205]: do we really need to add the return type?

I don't think we do that in a lot of places and we actually had a discussion a while ago where we decided to let TS infer the return type (at least for simple things like this)

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 7]
```diff
@@ -240,7 +268,7 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
         );
     }
 
-    renderUnreadScrollPositionLabel(option) {
+    renderUnreadScrollPositionLabel(option: string | undefined): JSX.Element {
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:271]:
```suggestion
    renderUnreadScrollPositionLabel(option?: string) {
```

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 8]
```diff
@@ -646,8 +674,8 @@ export default class AdvancedSettingsDisplay extends React.PureComponent {
             }
         }
 
-        let deactivateAccountSection = '';
-        let makeConfirmationModal = '';
+        let deactivateAccountSection: JSX.Element | string = '';
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:677]: IMO we could use `ReactNode` here as well

File: components/user_settings/user_settings.tsx [hunk 9]
```diff
@@ -16,7 +16,7 @@ export type Props = {
     user: UserProfile;
     activeTab?: string;
     activeSection: string;
-    updateSection: (section: string) => void;
+    updateSection: (section: string | undefined) => void;
```
Comment (michelengelen) [inline components/user_settings/user_settings.tsx:19]:
```suggestion
    updateSection: (section?: string) => void;
```

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 10]
```diff
@@ -1,57 +1,83 @@
...
-        enablePreviewFeatures: PropTypes.bool,
-        enableUserDeactivation: PropTypes.bool,
-        actions: PropTypes.shape({
-            savePreferences: PropTypes.func.isRequired,
-            updateUserActive: PropTypes.func.isRequired,
-            revokeAllSessionsForUser: PropTypes.func.isRequired,
-        }).isRequired,
-    }
+type Setting = {
+    [key: string]: string | undefined;
```
Comment (michelengelen) [inline components/user_settings/advanced/user_settings_advanced.tsx:31]: question: do we really have generic options we need to take into account?

**Review by michelengelen — CHANGES_REQUESTED** [review]
Just a few things to adjust and some nitpicking. :D

I have not added a comment on every param you typed as `string | undefined`, so could you please go over the changes and adjust all of them? Thanks!

#### Tip for the future
Use either optional parameters ...
```
const funcWithOptionalParam = (str?: string) => str;
```

... or default values and let TS infer the type ...

```
const funcWithDefaultValueParam = (str = 'some default string') => str;
```

**Review by hmhealey — APPROVED** [review]

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 6 above)
Comment (sinansonmez) [reply components/user_settings/advanced/user_settings_advanced.tsx:205]: In the previous PR reviews, I received a feedback to put return types for functions. For Future PRs and simple functions I can let TS info the return type.

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 10 above)
Comment (sinansonmez) [reply components/user_settings/advanced/user_settings_advanced.tsx:31]: Sorry, I didn't get the question 🥲

**Comment by sinansonmez** [comment]
Thanks for the comments @hmhealey Yes you are right. I always use `?` for type definitions for `Props` or `State` but for functions I forgot. Now it should all be fixed

**Review by michelengelen — APPROVED** [review]
approving to unblock here, but we should still be having that discussion about the return types @hmhealey ... it seems to be unclear which approach we do want to take on this.

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 6 above)
Comment (sinansonmez) [reply components/user_settings/advanced/user_settings_advanced.tsx:205]: Linking the discussion for everyone's visibility: https://github.com/mattermost/mattermost-webapp/pull/10661#discussion_r914934817

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 10 above)
Comment (sinansonmez) [reply components/user_settings/advanced/user_settings_advanced.tsx:31]: Thanks for the explanation. Now I see your point. 

The issue is that `as.name` can change in the future

https://github.com/mattermost/mattermost-webapp/blob/bd3fafaf1273fcf1ef593eb2a528545581eac3a1/components/user_settings/advanced/user_settings_advanced.tsx#L99-L100

Currently there is only one option (i.e. `feature_enabled_ markdown_preview` coming from below. If there will be more `PRE_RELEASE_FEATURES` in the future, I suppose it might create type issues

https://github.com/mattermost/mattermost-webapp/blob/bd3fafaf1273fcf1ef593eb2a528545581eac3a1/utils/constants.tsx#L1725-L1731

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 6 above)
Comment (hmhealey) [reply components/user_settings/advanced/user_settings_advanced.tsx:205]: I'll bring this up in the next Web Guild meeting to see if anyone has any strong opinions either way. Personally, I like them for functions that are being exported for use outside the web app since it means that we're less likely to accidentally introduce breaking changes, but inside the app, I don't particularly care one way or the other. Either way, we should try to get on the same page though to avoid confusing people during PR reviews with conflicting info

**Comment by hmhealey** [comment]
@jgilliam17 is out for the next week, and while I've assigned her for QA reviews so she sees this PR, I'll probably merge it if the E2E tests look good on it

**Comment by jgilliam17** [comment]
Thanks for running e2es @hmhealey 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/224088-7150fcc-onprem-ent-webapp-pr-10826-mattermost/mm-ee-test7150fcc/mochawesome.html) - looks good to merge, no PR related failures.
@sinansonmez thank you for your contribution 🙂

**Review by jgilliam17 — APPROVED** [review]
